src/router/sw_axi/router.go

```diff
@@ -47,17 +91,122 @@ func NewRouter(uri string, numClients int) (*Router, error) {
 	router := Router{}
 	router.uri = uri
 	router.numClients = numClients
-	router.ips = make(map[uint64]*IpInfo)
+	router.ipMMap = make(map[uint64]*IpInfo)
+	router.incoming = make(chan []byte)
 	return &router, nil
 }
 
-func (r *Router) allocateIp(ip *IpInfo) (uint64, error) {
+func (r *Router) registerIp(ip *IpInfo) (uint64, error) {
+	if ip.Type == MASTER || ip.Type == MASTER_LITE || ip.Type == MASTER_STREAM {
+		r.masterCount++
+	}
+
+	for _, rIp := range r.ips {
+		if ip.Address >= rIp.Address && ip.Address < (rIp.Address+rIp.Size) {
+			return 0, fmt.Errorf("Address space overlaps with already registered %s IP block", rIp.Name)
+		}
+		if ip.Address < rIp.Size && (ip.Address+ip.Size) > rIp.Address {
+			return 0, fmt.Errorf("Address space overlaps with already registered %s IP block", rIp.Name)
+		}
+	}
 	id := r.ipCount
+	ip.Id = id
 	r.ipCount++
-	r.ips[ip.Address] = ip
+	r.ipMMap[ip.Address] = ip
+	r.ips = append(r.ips, ip)
+	log.Infof("[%20s] %s", r.clients[ip.ClientId].SystemInfo.Name, ip.String())
 	return id, nil
 }
 
+func (r *Router) sendDone() {
+	builder := flatbuffers.NewBuilder(0)
+	wire.MessageStart(builder)
+	wire.MessageAddType(builder, wire.TypeDONE)
+	builder.Finish(wire.MessageEnd(builder))
+
+	for _, ch := range r.clients {
+		ch.outgoing <- builder.FinishedBytes()
+	}
+}
+
+// Return the IP ID and the Client ID of the IP block
+func (r *Router) findTarget(address, size uint64) (uint64, uint64, error) {
+	var ip *IpInfo
+	for _, rIp := range r.ipMMap {
+		if address >= rIp.Address && address < (rIp.Address+rIp.Size) {
+			ip = rIp
+			break
+		}
+	}
+
+	if ip == nil {
+		return 0, 0, fmt.Errorf("No IP block found at address: 0x%016x", address)
+	}
+
+	if address+size > ip.Address+ip.Size {
+		return 0, 0, fmt.Errorf("Transaction spans blocks bondary: %s", ip.Name)
+	}
+
+	return ip.Id, ip.ClientId, nil
+}
+
+func (r *Router) route() {
+	for {
+		msgArr := <-r.incoming
+		msg := wire.GetRootAsMessage(msgArr, 0)
+
+		if msg.Type() == wire.TypeTERMINATE {
+			ip := r.ips[msg.IpId()]
+			log.Infof("[%20s] %s terminated", r.clients[ip.ClientId].SystemInfo.Name, ip.Name)
+			r.masterCount--
+			if r.masterCount == 0 {
+				log.Infof("No active master remains")
+				r.sendDone()
+				break
+			}
+		}
+
+		if msg.Type() != wire.TypeTRANSACTION {
+			log.Fatalf("Received unexpected message: %s", msg.Type())
+		}
+
+		txn := msg.Txn(nil)
+		op := "Write"
+		if txn.Type() == wire.TransactionTypeREAD_RESP || txn.Type() == wire.TransactionTypeREAD_REQ {
+			op = "Read "
```

Comment:
This has a space at the end while "Write" has not. Not sure if this is significant.

Reply:
Yeah, the printout looks nicer when read and write (words) have equal width :)